src/builder.rs

```diff
@@ -0,0 +1,151 @@
+                    attrs: callback(AttributesBuilder { attrs: attrs }).into(),
+                }
+            }
+            Event::ClosedTag { name, attrs } => {
+                Event::ClosedTag {
+                    name: name,
+                    attrs: callback(AttributesBuilder { attrs: attrs }).into(),
+                }
+            }
+            _ => event,
+        }
+    }
+
+    pub fn set<N, V>(self, name: N, value: V) -> AttributesBuilder<'a>
+        where N: Into<Cow<'a, str>>,
+              V: Into<Cow<'a, str>>
+    {
+        self.modify(name, move |_| Some(value))
+    }
+
+    pub fn modify<N, V, F>(self, name: N, callback: F) -> AttributesBuilder<'a>
+        where N: Into<Cow<'a, str>>,
+              V: Into<Cow<'a, str>>,
+              F: FnOnce(Option<Cow<'a, str>>) -> Option<V>
+    {
+        let name = name.into();
+        let mut attrs = self.attrs.into_owned();
+        let index = attrs.iter().position(|attr| attr.name == name);
+        if let Some(value) = callback(index.map(|index| attrs.remove(index).value)) {
+            attrs.push(Attribute {
+                name: name,
+                value: value.into(),
+            });
+        }
+        AttributesBuilder { attrs: Cow::Owned(attrs) }
+    }
+}
+
+impl<'a> Into<Vec<Event<'a>>> for EventsBuilder<'a> {
+    fn into(self) -> Vec<Event<'a>> {
+        self.events
+    }
+}
+
+impl<'a> Into<AttributeSet<'a>> for AttributesBuilder<'a> {
+    fn into(self) -> AttributeSet<'a> {
+        self.attrs
+    }
+}
+
+impl<'a> From<[(&'a str, &'a str); 1]> for AttributesBuilder<'a> {
+    fn from(attrs: [(&'a str, &'a str); 1]) -> AttributesBuilder<'a> {
+        AttributesBuilder {
+            attrs: Cow::Owned(attrs.iter()
+                                   .map(|&(name, value)| {
+                                       Attribute {
+                                           name: name.into(),
+                                           value: value.into(),
+                                       }
+                                   })
+                                   .collect()),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::borrow::Borrow;
+    use super::{AttributesBuilder, EventsBuilder};
+    use Event;
+
+    fn prefix_ids(events: Vec<Event>, prefix: String) -> Vec<Event> {
+        events.into_iter()
+              .map(|event| {
+                  AttributesBuilder::rebuild(event, |attrs_builder| {
+                      attrs_builder.modify("id", |id| id.map(|id| prefix.clone() + id.borrow()))
+                  })
+              })
+              .collect()
+    }
+
+    #[test]
+    fn test_prefix() {
+        let mut events = EventsBuilder::new();
+
+        events.push_tag("h1", [("id", "hello")], |events| {
+            events.push_text("Hello, ");
+            events.push_tag("small", [("id", "world")], |events| {
+                events.push_text("world");
+            });
+        });
```

Comment:
This is kinda verbose to work with, and I feel that the internal code is not simple enough. As you said before about macros, I think this kind of builder pattern itself is best left for a different crate (which may provide it through macros).